### PR TITLE
Juann ferrari/refactor organizations controller

### PR DESCRIPTION
**Pull request by JuannFerrari**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

alt_text

### CODE REVIEW

File: app/controllers/organizations_controller.rb [hunk 1]
```diff
@@ -1,24 +1,15 @@
...
-    unless valid_image?
-      render template: "users/edit"
-      return
-    end
-
-    @organization = Organization.new(organization_params)
+    set_user_info
+    @form = OrganizationForm.new(organization_attributes: organization_params, current_user: current_user)
+    @organization = @form.organization
     authorize @organization
```
Comment (santib) [inline app/controllers/organizations_controller.rb:10]:
```suggestion
    authorize @form
```

and you'll need to add

```ruby
def model_name
  Organization.model_name
end
```
to the form object

File: app/controllers/organizations_controller.rb [hunk 2]
```diff
@@ -55,6 +39,19 @@ def generate_new_secret
 
   private
 
+  def handle_name_too_long
+    log_image_data_to_datadog
+  rescue Errno::ENAMETOOLONG
```
Comment (santib) [inline app/controllers/organizations_controller.rb:44]: I don't see why is this rescue needed, `log_image_data_to_datadog` shouldn't raise another error, right?

File: app/controllers/organizations_controller.rb [hunk 3]
```diff
@@ -95,39 +92,10 @@ def organization_params
   end
 
   def set_organization
-    @organization = Organization.find_by(id: organization_params[:id])
+    @organization = @form&.organization || Organization.find_by(id: organization_params[:id])
```
Comment (santib) [inline app/controllers/organizations_controller.rb:95]: when is `@form` or `@form.organization` be `nil`? My question is, when is the second part of the `||` going to be executed?

File: app/forms/organization_form.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    @organization_membership = OrganizationMembership.create!(organization: @organization,
+                                                              user: current_user,
+                                                              type_of_user: "admin")
+  end
+
+  private
+
+  def validate_image
+    return if valid_image_file?
```
Comment (santib) [inline app/forms/organization_form.rb]: don't you need to always validate the image filename too?

https://github.com/JuannFerrari/dev.to/pull/2/files#diff-33bef3083bae5b2e2bea316eab44c9fdL113

Comment (santib) [inline app/forms/organization_form.rb]: I think you need to `return if organization_attributes["profile_image"].blank?` wdyt?

https://github.com/JuannFerrari/dev.to/pull/2/files#diff-33bef3083bae5b2e2bea316eab44c9fdL106

File: app/controllers/organizations_controller.rb (hunk 2 above)
Comment (JuannFerrari) [reply app/controllers/organizations_controller.rb:44]: No, it always raise another error.

 ```
 def log_image_data_to_datadog
    images = Array.wrap(
      params.dig("user", "profile_image") ||
      params.dig("podcast", "image") ||
      params.dig("organization", "profile_image") ||
      params["image"],
    )

    raise if images.empty?

    images.each do |image|
      tags = [
        "controller:#{params['controller']}",
        "action:#{params['action']}",
        "content_type:#{image.content_type}",
        "original_filename:#{image.original_filename}",
        "tempfile:#{image.tempfile}",
        "size:#{image.size}",
      ]

      DatadogStatsClient.increment("image_upload_error", tags: tags)
    end

    raise
  end
```

File: spec/requests/organizations_create_spec.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    post "/organizations", params: { organization: organization.merge(profile_image: fixture_file_upload("files/800x600.png", "image/png")) }
+
+    tags = hash_including(tags: instance_of(Array))
+
+    expect(DatadogStatsClient).to have_received(:increment).with("image_upload_error", tags)
+  end
+
+  it "returns error if profile image file name is too long" do
+    form = OrganizationForm.new(organization_attributes: organization.merge(profile_updated_at: Time.current), current_user: user)
+    allow(OrganizationForm).to receive(:new).and_return(form)
```
Comment (santib) [inline spec/requests/organizations_create_spec.rb:37]: why mocking the form object initialization method?

File: app/controllers/organizations_controller.rb (hunk 3 above)
Comment (JuannFerrari) [reply app/controllers/organizations_controller.rb:95]: I think that when `generate_new_secret` is called. but I could change that method to use form object too.

**Pull request closed without merging**